File: aws/lambda/rss_api/codex_provider.py
from __future__ import annotations

import json
import os
import time
from typing import Any

import boto3
import requests


class CodexProviderError(RuntimeError):
    pass


class CodexAuthError(CodexProviderError):
    pass


class CodexSubscriptionProvider:
    BASE_URL = "https://chatgpt.com/backend-api/codex"
    REFRESH_URL = "https://auth.openai.com/oauth/token"
    CLIENT_ID = "app_EMoamEEZ73f0CkXaXp7hrann"
    RETRYABLE_STATUS_CODES = {429, 500, 502, 503, 504}

    def __init__(
        self,
        bucket: str | None = None,
        auth_key: str | None = None,
        model: str | None = None,
        client_version: str | None = None,
    ) -> None:
        self.bucket = bucket or os.environ["APP_BUCKET"]
        self.auth_key = auth_key or os.environ.get("CODEX_AUTH_S3_KEY", "codex/auth.json")
        self.base_url = os.environ.get("OPENAI_CODEX_BASE_URL", self.BASE_URL).rstrip("/")
        self.model = model or os.environ.get("OPENAI_CODEX_MODEL", "gpt-5.4")
        self.client_version = client_version or os.environ.get("OPENAI_CODEX_CLIENT_VERSION", "0.118.0")
        self.max_retries = int(os.environ.get("OPENAI_CODEX_MAX_RETRIES", "2"))
        self.retry_delay = float(os.environ.get("OPENAI_CODEX_RETRY_DELAY", "2.0"))
        self.s3 = boto3.client("s3")

    def configured(self) -> bool:
        try:
            self._load_auth_payload()
            return True
        except CodexAuthError:
            return False

    def _load_auth_payload(self) -> dict[str, Any]:
        try:
            response = self.s3.get_object(Bucket=self.bucket, Key=self.auth_key)
            payload = json.loads(response["Body"].read().decode("utf-8"))
        except Exception as exc:
            raise CodexAuthError(
                f"Codex auth JSON not configured at s3://{self.bucket}/{self.auth_key}"
            ) from exc
        tokens = payload.get("tokens") or {}
        if not tokens.get("access_token"):
            raise CodexAuthError("Codex auth JSON does not contain tokens.access_token")
        return payload

    def _persist_auth_payload(self, payload: dict[str, Any]) -> None:
        self.s3.put_object(
            Bucket=self.bucket,
            Key=self.auth_key,
            Body=json.dumps(payload).encode("utf-8"),
            ContentType="application/json",
            ServerSideEncryption="AES256",
        )

    def _refresh_access_token(self, payload: dict[str, Any]) -> None:
        refresh_token = (payload.get("tokens") or {}).get("refresh_token")
        if not refresh_token:
            raise CodexAuthError("Codex auth JSON does not contain a refresh token")
        response = requests.post(
            self.REFRESH_URL,
            headers={"Content-Type": "application/json"},
            json={
                "client_id": self.CLIENT_ID,
                "grant_type": "refresh_token",
                "refresh_token": refresh_token,
            },
            timeout=30,
        )
        if not response.ok:
            raise CodexAuthError(
                f"Failed to refresh Codex OAuth token: {response.status_code} {self._extract_error_message(response)}"
            )
        tokens = payload.setdefault("tokens", {})
        for key in ("id_token", "access_token", "refresh_token"):
            value = response.json().get(key)
            if value:
                tokens[key] = value
        payload["last_refresh"] = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        self._persist_auth_payload(payload)

    def _headers(self, payload: dict[str, Any]) -> dict[str, str]:
        token = (payload.get("tokens") or {}).get("access_token")
        return {"Authorization": f"Bearer {token}", "Content-Type": "application/json"}

    @staticmethod
    def _extract_error_message(response: requests.Response) -> str:
        try:
            payload = response.json()
        except ValueError:
            return response.text or "unknown error"
        if isinstance(payload, dict):
            detail = payload.get("detail")
            if detail:
                return str(detail)
            error = payload.get("error")
            if isinstance(error, dict):
                return str(error.get("message") or error.get("code") or "unknown error")
            if error:
                return str(error)
        return response.text or "unknown error"

    def _request(
        self,
        method: str,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        json_body: dict[str, Any] | None = None,
        timeout: int = 120,
        stream: bool = False,
    ) -> requests.Response:
        auth_payload = self._load_auth_payload()
        last_error: Exception | None = None
        for attempt in range(self.max_retries + 1):
            try:
                response = requests.request(
                    method,
                    f"{self.base_url}{path}",
                    headers=self._headers(auth_payload),
                    params=params,
                    json=json_body,
                    timeout=timeout,
                    stream=stream,
                )
            except requests.RequestException as exc:
                last_error = exc
                if attempt < self.max_retries:
                    time.sleep(self.retry_delay * (attempt + 1))
                    continue
                raise CodexProviderError(f"Codex request failed: {exc}") from exc

            if response.status_code == 401 and attempt < self.max_retries:
                self._refresh_access_token(auth_payload)
                auth_payload = self._load_auth_payload()
                continue
            if response.status_code in self.RETRYABLE_STATUS_CODES and attempt < self.max_retries:
                time.sleep(self.retry_delay * (attempt + 1))
                continue
            if response.status_code >= 400:
                message = self._extract_error_message(response)
                if response.status_code in (401, 403):
                    raise CodexAuthError(f"Codex auth failed: {response.status_code} {message}")
                raise CodexProviderError(f"Codex request failed: {response.status_code} {message}")
            return response
        if last_error:
            raise CodexProviderError(f"Codex request failed: {last_error}") from last_error
        raise CodexProviderError("Codex request failed without a response")

    def list_models(self) -> list[dict[str, Any]]:
        response = self._request(
            "GET",
            "/models",
            params={"client_version": self.client_version},
            timeout=30,
        )
        models = []
        for model in response.json().get("models", []):
            if not model.get("supported_in_api"):
                continue
            models.append(
                {
                    "id": model.get("slug"),
                    "label": model.get("display_name") or model.get("slug"),
                    "description": model.get("description"),
                    "visibility": model.get("visibility", "list"),
                    "priority": model.get("priority", 9999),
                    "defaultReasoningLevel": model.get("default_reasoning_level"),
                    "reasoningLevels": [
                        level.get("effort")
                        for level in model.get("supported_reasoning_levels", [])
                        if level.get("effort")
                    ],
                }
            )
        models.sort(key=lambda item: (item.get("visibility") != "list", item.get("priority", 9999), item.get("label") or ""))
        return models

    def generate(
        self,
        messages: list[dict[str, str]],
        *,
        model: str | None = None,
        max_tokens: int = 4000,
        reasoning_effort: str | None = None,
        timeout: int = 180,
    ) -> dict[str, Any]:
        used_model = model or self.model
        payload = self._messages_to_payload(messages, used_model, max_tokens, reasoning_effort)
        response = self._request("POST", "/responses", json_body=payload, timeout=timeout, stream=True)
        return self._parse_sse_response(response, used_model)

    def _messages_to_payload(
        self,
        messages: list[dict[str, str]],
        model: str,
        max_tokens: int,
        reasoning_effort: str | None,
    ) -> dict[str, Any]:
        instructions: list[str] = []
        normalized: list[dict[str, Any]] = []
        for message in messages:
            role = message.get("role", "user")
            content = message.get("content", "")
            if role == "system":
                if content:
                    instructions.append(content)
                continue
            normalized.append({"role": role, "content": [{"type": "input_text", "text": content}]})
        if not normalized:
            normalized.append({"role": "user", "content": [{"type": "input_text", "text": ""}]})
        payload: dict[str, Any] = {
            "model": model,
            "instructions": "\n\n".join(instructions) or "You are a helpful assistant.",
            "input": normalized,
            "parallel_tool_calls": False,
            "store": False,
            "stream": True,
        }
        verbosity = self._default_verbosity_for_max_tokens(max_tokens)
        if verbosity:
            payload["text"] = {"verbosity": verbosity}
        if reasoning_effort:
            payload["reasoning"] = {"effort": reasoning_effort}
        return payload

    @staticmethod
    def _default_verbosity_for_max_tokens(max_tokens: int) -> str | None:
        if max_tokens <= 1200:
            return "low"
        if max_tokens <= 2500:
            return "medium"
        return None

    def _parse_sse_response(self, response: requests.Response, model: str) -> dict[str, Any]:
        text_parts: list[str] = []
        completed: dict[str, Any] = {}
        for raw_line in response.iter_lines(decode_unicode=True):
            if isinstance(raw_line, bytes):
                raw_line = raw_line.decode("utf-8", errors="ignore")
            if not raw_line or not raw_line.startswith("data: "):
                continue
            try:
                event = json.loads(raw_line[6:])
            except json.JSONDecodeError:
                continue
            event_type = event.get("type")
            if event_type == "response.output_text.delta":
                text_parts.append(event.get("delta", ""))
            elif event_type == "response.output_text.done" and not text_parts:
                text_parts.append(event.get("text", ""))
            elif event_type == "response.completed":
                completed = event.get("response", {}) or {}
            elif event_type == "response.failed":
                error = (event.get("response") or {}).get("error") or {}
                raise CodexProviderError(error.get("message") or "Codex response failed")
        content = "".join(text_parts).strip()
        if not content:
            raise CodexProviderError("Codex returned an empty response body")
        usage = completed.get("usage") or {}
        return {
            "id": completed.get("id", f"codex-{int(time.time())}"),
            "model": completed.get("model", model),
            "content": content,
            "usage": {
                "promptTokens": usage.get("input_tokens"),
                "completionTokens": usage.get("output_tokens"),
                "totalTokens": usage.get("total_tokens"),
            },
        }

File: src/codex_provider.rs
use crate::{Result, RssAiError};
use reqwest::{Client, Response, StatusCode};
use serde_json::{json, Value};
use std::{env, path::PathBuf, process::Command, time::Duration};

const BASE_URL: &str = "https://chatgpt.com/backend-api/codex";
const REFRESH_URL: &str = "https://auth.openai.com/oauth/token";
const CLIENT_ID: &str = "app_EMoamEEZ73f0CkXaXp7hrann";
const DEFAULT_CLIENT_VERSION: &str = "0.118.0";
const RETRYABLE_STATUS_CODES: &[StatusCode] = &[
    StatusCode::TOO_MANY_REQUESTS,
    StatusCode::INTERNAL_SERVER_ERROR,
    StatusCode::BAD_GATEWAY,
    StatusCode::SERVICE_UNAVAILABLE,
    StatusCode::GATEWAY_TIMEOUT,
];

#[derive(Clone)]
pub struct CodexSubscriptionProvider {
    client: Client,
    auth_path: PathBuf,
    base_url: String,
    model: String,
    max_retries: usize,
    retry_delay: Duration,
}

impl CodexSubscriptionProvider {
    pub fn new(model: String) -> Self {
        Self {
            client: Client::new(),
            auth_path: auth_path(),
            base_url: env::var("OPENAI_CODEX_BASE_URL")
                .unwrap_or_else(|_| BASE_URL.to_string())
                .trim_end_matches('/')
                .to_string(),
            model,
            max_retries: parse_env("OPENAI_CODEX_MAX_RETRIES", 2),
            retry_delay: Duration::from_secs_f64(parse_env("OPENAI_CODEX_RETRY_DELAY", 2.0)),
        }
    }

    pub fn configured() -> bool {
        let path = auth_path();
        path.exists()
            && std::fs::read_to_string(path)
                .ok()
                .and_then(|content| serde_json::from_str::<Value>(&content).ok())
                .and_then(|payload| {
                    payload
                        .get("tokens")
                        .and_then(|tokens| tokens.get("access_token"))
                        .and_then(Value::as_str)
                        .map(|token| !token.is_empty())
                })
                .unwrap_or(false)
    }

    pub async fn generate(
        &self,
        messages: Vec<ChatMessage>,
        max_tokens: usize,
        reasoning_effort: Option<String>,
    ) -> Result<String> {
        let payload = self.messages_to_payload(messages, max_tokens, reasoning_effort);
        let response = self
            .request_with_retries("POST", "/responses", Some(payload), 180)
            .await?;
        let body = response.text().await?;
        parse_sse_response(&body)
    }

    fn messages_to_payload(
        &self,
        messages: Vec<ChatMessage>,
        max_tokens: usize,
        reasoning_effort: Option<String>,
    ) -> Value {
        let mut instructions = Vec::new();
        let mut input = Vec::new();

        for message in messages {
            if message.role == "system" {
                if !message.content.is_empty() {
                    instructions.push(message.content);
                }
                continue;
            }

            input.push(json!({
                "role": message.role,
                "content": [{"type": "input_text", "text": message.content}],
            }));
        }

        if input.is_empty() {
            input.push(json!({
                "role": "user",
                "content": [{"type": "input_text", "text": ""}],
            }));
        }

        let mut payload = json!({
            "model": self.model,
            "instructions": if instructions.is_empty() {
                "You are a helpful assistant.".to_string()
            } else {
                instructions.join("\n\n")
            },
            "input": input,
            "parallel_tool_calls": false,
            "store": false,
            "stream": true,
        });

        if let Some(verbosity) = default_verbosity_for_max_tokens(max_tokens) {
            payload["text"] = json!({ "verbosity": verbosity });
        }

        if let Some(effort) = reasoning_effort.filter(|value| !value.is_empty()) {
            payload["reasoning"] = json!({ "effort": effort });
        }

        payload
    }

    async fn request_with_retries(
        &self,
        method: &str,
        path: &str,
        json_body: Option<Value>,
        timeout_secs: u64,
    ) -> Result<Response> {
        let mut auth_payload = self.load_auth_payload()?;
        let mut last_error = None;

        for attempt in 0..=self.max_retries {
            let response = self
                .send_request(method, path, json_body.clone(), &auth_payload, timeout_secs)
                .await;

            let response = match response {
                Ok(response) => response,
                Err(err) => {
                    last_error = Some(err.to_string());
                    if attempt < self.max_retries {
                        self.sleep_before_retry(attempt).await;
                        continue;
                    }
                    return Err(err.into());
                }
            };

            if response.status() == StatusCode::UNAUTHORIZED && attempt < self.max_retries {
                self.refresh_access_token(&mut auth_payload).await?;
                auth_payload = self.load_auth_payload()?;
                continue;
            }

            if RETRYABLE_STATUS_CODES.contains(&response.status()) && attempt < self.max_retries {
                self.sleep_before_retry(attempt).await;
                continue;
            }

            if response.status().is_client_error() || response.status().is_server_error() {
                let status = response.status();
                let message = extract_error_message_from_response(response).await;
                return Err(RssAiError::AiApi(format!(
                    "Codex request failed: {} {}",
                    status, message
                )));
            }

            return Ok(response);
        }

        Err(RssAiError::AiApi(format!(
            "Codex request failed{}",
            last_error
                .map(|err| format!(": {}", err))
                .unwrap_or_else(|| " without a response".to_string())
        )))
    }

    async fn send_request(
        &self,
        method: &str,
        path: &str,
        json_body: Option<Value>,
        auth_payload: &Value,
        timeout_secs: u64,
    ) -> reqwest::Result<Response> {
        let token = auth_payload
            .get("tokens")
            .and_then(|tokens| tokens.get("access_token"))
            .and_then(Value::as_str)
            .unwrap_or("");

        let request = match method {
            "POST" => self.client.post(format!("{}{}", self.base_url, path)),
            "GET" => self.client.get(format!("{}{}", self.base_url, path)),
            _ => self.client.request(
                method.parse().unwrap_or(reqwest::Method::GET),
                format!("{}{}", self.base_url, path),
            ),
        }
        .bearer_auth(token)
        .header("Content-Type", "application/json")
        .timeout(Duration::from_secs(timeout_secs));

        if let Some(body) = json_body {
            request.json(&body).send().await
        } else {
            request.send().await
        }
    }

    async fn refresh_access_token(&self, auth_payload: &mut Value) -> Result<()> {
        let refresh_token = auth_payload
            .get("tokens")
            .and_then(|tokens| tokens.get("refresh_token"))
            .and_then(Value::as_str)
            .ok_or_else(|| {
                RssAiError::AiApi(
                    "Codex auth payload does not contain a refresh token. Run `codex login` again."
                        .to_string(),
                )
            })?;

        let response = self
            .client
            .post(REFRESH_URL)
            .header("Content-Type", "application/json")
            .timeout(Duration::from_secs(30))
            .json(&json!({
                "client_id": CLIENT_ID,
                "grant_type": "refresh_token",
                "refresh_token": refresh_token,
            }))
            .send()
            .await?;

        if !response.status().is_success() {
            let status = response.status();
            let message = extract_error_message_from_response(response).await;
            return Err(RssAiError::AiApi(format!(
                "Failed to refresh Codex OAuth token: {} {}",
                status, message
            )));
        }

        let refreshed: Value = response.json().await?;
        let tokens = auth_payload
            .get_mut("tokens")
            .and_then(Value::as_object_mut)
            .ok_or_else(|| RssAiError::AiApi("Codex auth payload is missing tokens".to_string()))?;

        for key in ["id_token", "access_token", "refresh_token"] {
            if let Some(value) = refreshed.get(key).cloned().filter(|value| !value.is_null()) {
                tokens.insert(key.to_string(), value);
            }
        }

        auth_payload["last_refresh"] = Value::String(chrono::Utc::now().to_rfc3339());
        self.persist_auth_payload(auth_payload)
    }

    fn load_auth_payload(&self) -> Result<Value> {
        let content = std::fs::read_to_string(&self.auth_path).map_err(|err| {
            RssAiError::AiApi(format!(
                "Codex auth file not found at {}: {}. Run `codex login` first.",
                self.auth_path.display(),
                err
            ))
        })?;

        let payload: Value = serde_json::from_str(&content).map_err(|err| {
            RssAiError::AiApi(format!("Failed to parse Codex auth JSON: {}", err))
        })?;

        let has_access_token = payload
            .get("tokens")
            .and_then(|tokens| tokens.get("access_token"))
            .and_then(Value::as_str)
            .map(|token| !token.is_empty())
            .unwrap_or(false);

        if !has_access_token {
            return Err(RssAiError::AiApi(
                "Codex auth file does not contain tokens.access_token".to_string(),
            ));
        }

        Ok(payload)
    }

    fn persist_auth_payload(&self, payload: &Value) -> Result<()> {
        let content = serde_json::to_string_pretty(payload).map_err(|err| {
            RssAiError::AiApi(format!("Failed to serialize Codex auth JSON: {}", err))
        })?;
        std::fs::write(&self.auth_path, content)?;
        Ok(())
    }

    async fn sleep_before_retry(&self, attempt: usize) {
        tokio::time::sleep(self.retry_delay.mul_f64((attempt + 1) as f64)).await;
    }
}

#[derive(Clone)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
}

pub fn default_codex_model() -> String {
    env::var("OPENAI_CODEX_MODEL").unwrap_or_else(|_| "gpt-5.4".to_string())
}

pub fn detect_client_version() -> String {
    env::var("OPENAI_CODEX_CLIENT_VERSION").unwrap_or_else(|_| {
        Command::new("codex")
            .arg("--version")
            .output()
            .ok()
            .and_then(|output| {
                let stdout = String::from_utf8_lossy(&output.stdout);
                stdout.split_whitespace().nth(1).map(str::to_string)
            })
            .unwrap_or_else(|| DEFAULT_CLIENT_VERSION.to_string())
    })
}

fn auth_path() -> PathBuf {
    if let Ok(path) = env::var("OPENAI_CODEX_AUTH_PATH") {
        return expand_home(path);
    }

    env::var("HOME")
        .map(|home| PathBuf::from(home).join(".codex").join("auth.json"))
        .unwrap_or_else(|_| PathBuf::from(".codex/auth.json"))
}

fn expand_home(path: String) -> PathBuf {
    if let Some(rest) = path.strip_prefix("~/") {
        if let Ok(home) = env::var("HOME") {
            return PathBuf::from(home).join(rest);
        }
    }
    PathBuf::from(path)
}

fn parse_env<T>(key: &str, default: T) -> T
where
    T: std::str::FromStr,
{
    env::var(key)
        .ok()
        .and_then(|value| value.parse().ok())
        .unwrap_or(default)
}

fn default_verbosity_for_max_tokens(max_tokens: usize) -> Option<&'static str> {
    if max_tokens <= 1200 {
        Some("low")
    } else if max_tokens <= 2500 {
        Some("medium")
    } else {
        None
    }
}

async fn extract_error_message_from_response(response: Response) -> String {
    let text = response
        .text()
        .await
        .unwrap_or_else(|_| "unknown error".to_string());
    extract_error_message(&text)
}

fn extract_error_message(text: &str) -> String {
    let Ok(payload) = serde_json::from_str::<Value>(text) else {
        return if text.is_empty() {
            "unknown error".to_string()
        } else {
            text.to_string()
        };
    };

    if let Some(detail) = payload.get("detail").and_then(Value::as_str) {
        return detail.to_string();
    }

    if let Some(error) = payload.get("error") {
        if let Some(message) = error.get("message").and_then(Value::as_str) {
            return message.to_string();
        }
        if let Some(code) = error.get("code").and_then(Value::as_str) {
            return code.to_string();
        }
        if let Some(message) = error.as_str() {
            return message.to_string();
        }
    }

    text.to_string()
}

fn parse_sse_response(body: &str) -> Result<String> {
    let mut text_parts = Vec::new();

    for line in body.lines() {
        let Some(data) = line.strip_prefix("data: ") else {
            continue;
        };
        if data == "[DONE]" {
            continue;
        }

        let Ok(event) = serde_json::from_str::<Value>(data) else {
            continue;
        };

        match event.get("type").and_then(Value::as_str) {
            Some("response.output_text.delta") => {
                if let Some(delta) = event.get("delta").and_then(Value::as_str) {
                    text_parts.push(delta.to_string());
                }
            }
            Some("response.output_text.done") if text_parts.is_empty() => {
                if let Some(text) = event.get("text").and_then(Value::as_str) {
                    text_parts.push(text.to_string());
                }
            }
            Some("response.failed") => {
                let message = event
                    .get("response")
                    .and_then(|response| response.get("error"))
                    .and_then(|error| error.get("message"))
                    .and_then(Value::as_str)
                    .unwrap_or("Codex response failed");
                return Err(RssAiError::AiApi(message.to_string()));
            }
            _ => {}
        }
    }

    let content = text_parts.join("").trim().to_string();
    if content.is_empty() {
        return Err(RssAiError::AiApi(
            "Codex returned an empty response body".to_string(),
        ));
    }

    Ok(content)
}
